### ISSUE TITLE
[feature request] Storing device specific values on NEEO.

### ISSUE THREAD

**Issue by nklerk** [NEEOInc/neeo-sdk#99]
Everytime I start a driver i first have to do a discovery. this in itself isn't an issue but in some cases it could help enormously if more values could be stored besides the unique id. An example i would like to store and be able to get using the SDK:

MAC address,  //Using this currently as deviceId. Important for the WOL feature of the driver.
Hostname,  //Using this to connect to the device without the need of additional discovery methods.
Password, //Storing password so that when the driver is killed you don't have to re add the device every time.
User settings. things like filter preference, what lists to show or not to show, thus providing some customizability through the driver.

**Comment by pfiaux** [comment]
We are considering adding some helpers for storage to the SDK, would that cover those needs?

The settings you listed look specific to the SDK device itself and not data that would be needed on the Brain.

I think it makes more sense for the SDK to be able to save/load it's own values than to push them too the Brain. It keeps the 2 more decoupled.

**Comment by nklerk** [comment]
I believe that would cover most of my needs. 

In my opinion storing local or storing on brain are both viable options. especially in a multi brain configuration like i have, parameters could differ per neeo combo while using the same driver. maybe I'm looking to far in the future as the SDK is currently not multi brain. Please forgive my enthusiasm ;-)

A good example is my kodi setup. I have one HTPC centralized in our home and I'm distributing the HDMI with a HD Anywhere MHUB to multiple TV's. every room with a TV also has a NEEO Combo. The living room would have different settings then the children's room. sure this is an edge case but it's easyer to share the things you like to solve than come up with other good examples.

**Comment by carp3-noctem** [comment]
I agree, that normaly it would make sense to store Infos in the Driver itself, but such things as Username and Password can be changed and needs to be updated from side of the user, without running through the setup again. Therefor it makes sense to use the Brain as Storage and also provide the ability to update these Informations. This is also half way implemented in the Brain, so it just needs the ability to overhand these Infos to the SDK.
Also makes sense to be able to check, if these Infos are allready implementes (in case of Spotify for example) if allready entered it would be great to use it for Spotify Connect drivers or other services.
OF COURSE the access to these Credentials needs to be allowed by the User through the Setup as well as when there was a change in the Credentials.

**Comment by neophob** [comment]
So to make this a bit more actionable I think we should focus on the driver specific implementation and exclude data stored and have a managed lifecycle (user can update, delete his account information for example).

So your request @nklerk was to store *some* data for the driver. So then it must be in the responsibility of the driver to maintain this data - so saving this data on the Brain is not a good idea.

To summarise:

- you would like an option from the sdk to store some generic data (like last used playlist as example) which are NOT needed to run you driver.

**Comment by neophob** [comment]
@nklerk any input from you on this issue? the latest SDK supports a "security code" feature (the user can add enter security code - see ikea tradfri). we will implement a username/password feature in a future version too.

about storing data - the brain stored the device id the driver specify - but the rest should be stored by the driver itself.

**Comment by nklerk** [comment]
The reason I request this feature is because i'm used to have this as a feature in another HA product. I can store values and read values in a structured way and every "driver" uses the same method. the SDK is in that case helping out people like me.

what I had in mind is something like this:


```
const sdk = require "neeosdk";

sdk.configuration.store('shared secret', 'Secr3T');

let secret = sdk.configuration.read('shared secret')
```



or 

```
const sdk = require "neeosdk";

sdk.configuration.store('preferances', {a:"A",b:"B"});
let 'preferances'= sdk.configuration.read('preferances')
```

**Comment by pfiaux** [comment]
Since we're talking about doing the storage on the SDK side, I think it would make sense to do this via the https://github.com/NEEOInc/neeo-sdk-toolkit/tree/master/ repo via a separate package.

The advantage is it's decoupled from the SDK core and it's not tied to the Brain firmware versions and can be updated more frequently.

Perhaps we can reuse part of https://github.com/NEEOInc/neeo-sdk/pull/80 for that.

**Comment by pfiaux** [comment]
@nklerk I would recommend using https://www.npmjs.com/package/conf

I don't think we need a NEEO-SDK specific implementation yet.

**Comment by nklerk** [comment]
Thanks!
